llvm/lib/Target/AMDGPU/AMDGPULowerKernelAttributes.cpp

```diff
@@ -317,10 +317,38 @@ static bool processUse(CallInst *CI, bool IsV5OrAbove) {
   return MadeChange;
 }
 
+// SYCL allows required work-group size attribute to be partially specified
+// (not all three dimensions), provide a default value (1) for the missing
+// dimensions.
+static void updateSYCLreqdWorkGroupMD(Function &F) {
+  auto *Node = F.getMetadata("reqd_work_group_size");
+  if (!Node || Node->getNumOperands() == 3)
+    return;
+
+  auto &Context = F.getContext();
+  SmallVector<uint64_t, 3> RWGS;
+  for (auto &Op : Node->operands())
+    RWGS.push_back(mdconst::extract<ConstantInt>(Op)->getZExtValue());
+  while (RWGS.size() != 3)
+    RWGS.push_back(1);
+
+  llvm::Metadata *RWGSArgs[] = {
+      llvm::ConstantAsMetadata::get(llvm::ConstantInt::get(
+          llvm::IntegerType::get(Context, 32), llvm::APInt(32, RWGS[0]))),
+      llvm::ConstantAsMetadata::get(llvm::ConstantInt::get(
+          llvm::IntegerType::get(Context, 32), llvm::APInt(32, RWGS[1]))),
+      llvm::ConstantAsMetadata::get(llvm::ConstantInt::get(
+          llvm::IntegerType::get(Context, 32), llvm::APInt(32, RWGS[2])))};
+  F.setMetadata("reqd_work_group_size", llvm::MDNode::get(Context, RWGSArgs));
+}
 
 // TODO: Move makeLIDRangeMetadata usage into here. Seem to not get
 // TargetPassConfig for subtarget.
 bool AMDGPULowerKernelAttributes::runOnModule(Module &M) {
+  for (auto &F : M)
+    if (F.hasFnAttribute("sycl-module-id"))
```

Comment:
This is unnecessary

Reply:
This is an error condition in OpenCL, which is check by the verifier: https://github.com/llvm/llvm-project/blob/856bd99aaedb295a4c9f221a8073e4f67c209d3a/llvm/lib/BinaryFormat/AMDGPUMetadataVerifier.cpp#L232
However, it is acceptable in SYCL.

Reply:
What's an error condition? That verifier is just checking the final output parses? The backend shouldn't need language specific checks

Reply:
Is it just tripping over the assert 3 element condition? But you fixed that by adjusting the metadata to match?

Reply:
Sorry, I should have been more clear.

> What's an error condition? That verifier is just checking the final output parses?

No, OpenCL mandates all 3 dimensions are set, the verifier, rightly so, expects 3 elements and errors out on sub 3 elements. It's just that the same metadata now can be created through either SYCL, or OpenCL, and SYCL happens to relax the restriction.

> But you fixed that by adjusting the metadata to match?

Yes, with this patch the condition is also true for SYCL's required work group size (padded with 1 for sub 3 elements).

A quick grep through ROCm-Developer-Tools, shows that this assumption (all 3 elements specified) is relied upon in a bunch of places: 
https://github.com/ROCm-Developer-Tools/clr/blob/38d2c56784fe2a2b9aff35822d3c9f4616189ead/rocclr/device/devkernel.cpp#L216
https://github.com/ROCm-Developer-Tools/clr/blob/38d2c56784fe2a2b9aff35822d3c9f4616189ead/rocclr/device/devkernel.cpp#L250

Reply:
@arsenm  did I make a convincing argument for this patch, could we please procede?

Reply:
Wouldn't `updateSYCLreqdWorkGroupMD` just early exit for OpenCL anyway since it always has 3 elements:

```cpp
  auto *Node = F.getMetadata("reqd_work_group_size");
  if (!Node || Node->getNumOperands() == 3)
    return;
```

So we should be able to skip this `sycl-module-id` check and just always call the attribute updating function

Reply:
Yeap, it would, but I wanted to be explicit about the fact that this is a sycl specific update. Happy to get rid of it though.